test/known_problems/should_pass/operator_subtypes.erl

```diff
@@ -0,0 +1,22 @@
+-module(operator_subtypes).
+-compile([pos_op1/2, nonneg_op1/2, neg_op1/2, word_op1/1, word_op2/2, word_op3/3]).
+
+-type word16() :: 0..65535.
+
+-spec pos_op1(pos_integer(), 1..10) -> pos_integer().
+pos_op1(X, Y) -> X * Y + 777 bor Y.
```

Comment:
Clauses for `pos_integer()` and `non_neg_integer()` could be added to `+` and `*`, just so we don't break existing rules where it's possible to avoid it.

The special rules for bitwise operations closed under types on the form 0..2<sup>n</sup>-1 aren't possible to express in the type language syntax, so I suggest you keep Ulf's code for now.

Reply:
For the record, e.g. the type we have for `bxor` is equivallent to the inifite clause spec

```Erlang

-spec erlang:'bxor'(0..16#1, 0..16#1) -> 0..16#1; % 1 bit
-spec erlang:'bxor'(0..16#3, 0..16#3) -> 0..16#3; % 2 bits
-spec erlang:'bxor'(0..16#7, 0..16#7) -> 0..16#7; % 3 bits
-spec erlang:'bxor'(0..16#f, 0..16#f) -> 0..16#f; % 4 bits

-spec erlang:'bxor'(0..16#1f, 0..16#1f) -> 0..16#1f;
-spec erlang:'bxor'(0..16#3f, 0..16#3f) -> 0..16#3f;
-spec erlang:'bxor'(0..16#7f, 0..16#7f) -> 0..16#7f;
-spec erlang:'bxor'(0..16#ff, 0..16#ff) -> 0..16#ff; % 8 bits

-spec erlang:'bxor'(0..16#1ff, 0..16#1ff) -> 0..16#1ff;
-spec erlang:'bxor'(0..16#3ff, 0..16#3ff) -> 0..16#3ff;
-spec erlang:'bxor'(0..16#7ff, 0..16#7ff) -> 0..16#7ff;
-spec erlang:'bxor'(0..16#fff, 0..16#fff) -> 0..16#fff;

-spec erlang:'bxor'(0..16#1fff, 0..16#1fff) -> 0..16#1fff;
-spec erlang:'bxor'(0..16#3fff, 0..16#3fff) -> 0..16#3fff;
-spec erlang:'bxor'(0..16#7fff, 0..16#7fff) -> 0..16#7fff;
-spec erlang:'bxor'(0..16#ffff, 0..16#ffff) -> 0..16#ffff; % 16 bits

-spec erlang:'bxor'(0..16#1ffff, 0..16#1ffff) -> 0..16#1ffff;
-spec erlang:'bxor'(0..16#3ffff, 0..16#3ffff) -> 0..16#3ffff;
-spec erlang:'bxor'(0..16#6ffff, 0..16#7ffff) -> 0..16#7ffff;
-spec erlang:'bxor'(0..16#7ffff, 0..16#fffff) -> 0..16#fffff;

-spec erlang:'bxor'(0..16#1fffff, 0..16#1fffff) -> 0..16#1fffff;
-spec erlang:'bxor'(0..16#3fffff, 0..16#3fffff) -> 0..16#3fffff;
-spec erlang:'bxor'(0..16#6fffff, 0..16#7fffff) -> 0..16#7fffff;
-spec erlang:'bxor'(0..16#7fffff, 0..16#ffffff) -> 0..16#ffffff;

-spec erlang:'bxor'(0..16#1ffffff, 0..16#1ffffff) -> 0..16#1ffffff;
-spec erlang:'bxor'(0..16#3ffffff, 0..16#3ffffff) -> 0..16#3ffffff;
-spec erlang:'bxor'(0..16#6ffffff, 0..16#7ffffff) -> 0..16#7ffffff;
-spec erlang:'bxor'(0..16#7ffffff, 0..16#fffffff) -> 0..16#fffffff;

-spec erlang:'bxor'(0..16#1fffffff, 0..16#1fffffff) -> 0..16#1fffffff;
-spec erlang:'bxor'(0..16#3fffffff, 0..16#3fffffff) -> 0..16#3fffffff;
-spec erlang:'bxor'(0..16#6fffffff, 0..16#7fffffff) -> 0..16#7fffffff;
-spec erlang:'bxor'(0..16#7fffffff, 0..16#ffffffff) -> 0..16#ffffffff; % 32 bits

%% and so on....
```

Reply:
Maybe, if we want to draw a line somewhere, it's enough if we keep only ~10 of these clauses for only the powers-of-two bits (1, 2, 4, 8, 16, 32, 64, 128, 256). That's good enough IMO, if it means we can stick to the -type/-spec notation in general.

For the record # 2: We mentioned that we'll solve constraint solving before changing this, to make sure these specs don't explode performance-wise.